app/src/main/java/org/oppia/app/utility/Logger.kt

```diff
@@ -0,0 +1,126 @@
+  private val LOG_DIRECTORY = File(context.filesDir, "oppia_app.log")
+
+  private enum class LogLevel private constructor(val logLevel: Int) {
+    VERBOSE(Log.VERBOSE),
+    DEBUG(Log.DEBUG),
+    INFO(Log.INFO),
+    WARNING(Log.WARN),
+    ERROR(Log.ERROR),
+    ASSERT(Log.ASSERT)
+  }
+
+  /** Logs a verbose message with the specified tag.*/
+  fun v(tag: String, msg: String) {
+    write(LogLevel.VERBOSE, tag, msg)
+  }
+
+  /** Logs a verbose message with the specified tag, message and exception.*/
+  fun v(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.VERBOSE, tag, msg, tr)
+  }
+
+  /** Logs a debug message with the specified tag*/
+  fun d(tag: String, msg: String) {
+    write(LogLevel.DEBUG, tag, msg)
+  }
+
+  /** Logs a debug message with the specified tag, message and exception.*/
+  fun d(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.DEBUG, tag, msg, tr)
+  }
+
+  /** Logs a info message with the specified tag.*/
+  fun i(tag: String, msg: String) {
+    write(LogLevel.INFO, tag, msg)
+  }
+
+  /** Logs a info message with the specified tag, message and exception.*/
+  fun i(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.INFO, tag, msg, tr)
+  }
+
+  /** Logs a warn message with the specified tag.*/
+  fun w(tag: String, msg: String) {
+    write(LogLevel.WARNING, tag, msg)
+  }
+
+  /** Logs a warn message with the specified tag, message and exception.*/
+  fun w(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.WARNING, tag, msg, tr)
+  }
+
+  /** Logs a error message with the specified tag.*/
+  fun e(tag: String, msg: String) {
+    write(LogLevel.ERROR, tag, msg)
+  }
+
+  /** Logs a error message with the specified tag, message and exception.*/
+  fun e(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.ERROR, tag, msg, tr)
+  }
+
+  private fun isLogEnable(logLevel: LogLevel): Boolean {
+    return GLOBAL_LOG_LEVEL.logLevel < logLevel.logLevel
+  }
+
+  private fun write(logLevel: LogLevel, tag: String, log: String) {
+    if (isLogEnable(logLevel) && ENABLE_CONSOLE_LOG) {
+      Log.println(logLevel.logLevel, tag, log)
+    }
+    if (isLogEnable(logLevel) && ENABLE_FILE_LOG) {
+      val msg = Calendar.getInstance().time.toString() + "\t" + logLevel.name + "/" + tag + ": " + log
+      /** To mange background threads*/
+      CoroutineScope(IO).launch { write(msg) }
+    }
+  }
+
+  private fun write(logLevel: LogLevel, tag: String, log: String, tr: Throwable) {
+    if (isLogEnable(logLevel) && ENABLE_CONSOLE_LOG) {
+      Log.println(logLevel.logLevel, tag, log + "\n" + Log.getStackTraceString(tr))
+    }
+    if (isLogEnable(logLevel) && ENABLE_FILE_LOG) {
+      val msg =
+        Calendar.getInstance().time.toString() + "\t" + logLevel.name + "/" + tag + ": " + log + "\n" + Log.getStackTraceString(
+          tr
+        )
+      /** To mange background threads*/
+     CoroutineScope(IO).launch { write(msg) }
```

Comment:
I think we actually want to inject the blocking dispatcher like we use in ``InMemoryCache`` instead of using the I/O dispatcher. This is definitely an interesting dispatcher that I wasn't aware of, but per https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-i-o.html there are a lot of potential threads this can run on. That could result in log messages being written out-of-order. To minimize this risk, I think we want to use a single thread to perform all writing, just not the main thread.

Reply:
You want me to run its own single thread?

Reply:
Something like this which will run on its own thread  `CoroutineScope(newSingleThreadContext("MyOwnThread")).launch { write(msg) }`